gateway/gateway_test.go

```diff
@@ -504,317 +116,614 @@ func TestPretty404(t *testing.T) {
 	} {
 		testName := fmt.Sprintf("%s %s", test.path, test.accept)
 		t.Run(testName, func(t *testing.T) {
-			var c http.Client
-			req, err := http.NewRequest("GET", ts.URL+test.path, nil)
-			assert.NoError(t, err)
+			req := mustNewRequest(t, "GET", ts.URL+test.path, nil)
 			req.Header.Add("Accept", test.accept)
 			req.Host = host
-			resp, err := c.Do(req)
-			assert.NoError(t, err)
+			resp := mustDo(t, req)
 			defer resp.Body.Close()
-			assert.Equal(t, test.status, resp.StatusCode)
+			require.Equal(t, test.status, resp.StatusCode)
 			body, err := io.ReadAll(resp.Body)
-			assert.NoError(t, err)
+			require.NoError(t, err)
 			if test.text != "" {
-				assert.Equal(t, test.text, string(body))
+				require.Equal(t, test.text, string(body))
 			}
 		})
 	}
 }
 
-func TestBrowserErrorHTML(t *testing.T) {
+func TestHeaders(t *testing.T) {
+	t.Parallel()
+
 	ts, _, root := newTestServerAndNode(t, nil)
-	t.Logf("test server url: %s", ts.URL)
 
-	t.Run("plain error if request does not have Accept: text/html", func(t *testing.T) {
-		t.Parallel()
+	var (
+		dirCID   = "bafybeihta5xfgxcmyxyq6druvidc7es6ogffdd6zel22l3y4wddju5xxsu"
+		dirPath  = "/ipfs/" + root.String() + "/subdir/"
+		dirRoots = "bafybeifhvgr4ufgwpoj2iyrymrigjhpexvsyfm2elafebkmrei4skunihe," + dirCID
+
+		fileCID   = "bafkreiba3vpkcqpc6xtp3hsatzcod6iwneouzjoq7ymy4m2js6gc3czt6i"
+		filePath  = "/ipfs/" + root.String() + "/subdir/fnord"
+		fileRoots = dirRoots + "," + fileCID
+
+		dagCborCID   = "bafyreiaocls5bt2ha5vszv5pwz34zzcdf3axk3uqa56bgsgvlkbezw67hq"
+		dagCborPath  = "/ipfs/" + root.String() + "/subdir/dag-cbor-document"
+		dagCborRoots = dirRoots + "," + dagCborCID
```

Comment:
We need to add `X-Ipfs-Roots` test for HAMT.
We should have test for `/ipfs/cid/hamt/file` and only have logical roots (one per path segment, and not all internal HAMT ones)